### PR TITLE
Support string literal bounds and return types on Config#get and Config#require

### PR DESCRIPTION
**Pull request by AaronFriel**
See attached image for an example of the impact:

image

This enables exhaustiveness checking in switch statements, code completion and inference on checking the return value, etc.

If `allowedValues` is a `string[]`, then the constraint `K` resolves to `string`. Should be backwards compatible.

I signed off this commit as a DCO and would be willing to sign a CLA.

### CODE REVIEW

File: sdk/nodejs/config.ts [hunk 1]
```diff
@@ -49,13 +49,14 @@ export class Config {
...
-    public get(key: string, opts?: StringConfigOptions): string | undefined {
-        const v = getConfig(this.fullKey(key));
+    public get<K extends string = string>(key: string, opts?: StringConfigOptions<K>): K | undefined {
+        const v: string | undefined = getConfig(this.fullKey(key));
         if (v === undefined) {
             return undefined;
         }
         if (opts) {
-            if (opts.allowedValues !== undefined && opts.allowedValues.indexOf(v) === -1) {
+            // SAFETY: if allowedValues != null, verifying v ∈ K[]
```
Comment (CyrusNajmabadi) [inline sdk/nodejs/config.ts]: i'm not sure that this comment adds any value.  it seems to just be restating the code below.

File: sdk/nodejs/config.ts [hunk 2]
```diff
@@ -71,7 +72,10 @@ export class Config {
                 }
             }
         }
-        return v;
+        // SAFETY:
+        // allowedValues != null ⇒ v ∈ K[]
+        // allowedValues == null ⇒ K = string & v : string
```
Comment (CyrusNajmabadi) [inline sdk/nodejs/config.ts]: this doesn't really follow our style on comments.  We don't really try to use these sorts of mathematical invariants in comments.

**Comment by CyrusNajmabadi** [comment]
Nice.  Thanks for the PR!

File: sdk/nodejs/config.ts (hunk 1 above)
Comment (AaronFriel) [reply sdk/nodejs/config.ts]: I tend to write `SAFETY` comments when performing potentially unsafe casts (`v as K`). Can remove.

File: sdk/nodejs/config.ts (hunk 2 above)
Comment (AaronFriel) [reply sdk/nodejs/config.ts]: I tend to write `SAFETY` comments when performing potentially unsafe casts (`v as K`). Can remove.

File: sdk/nodejs/config.ts (hunk 1 above)
Comment (CyrusNajmabadi) [reply sdk/nodejs/config.ts]: I'm ok with that.  We can take this as is. :)

**Comment by CyrusNajmabadi** [comment]
Thanks for the PR!

**Pull request merged**